### PR TITLE
Use ruff for linting (instead of current flake8/isort/pyflakes)

### PR DESCRIPTION
**Pull request by jairhenrique**
https://github.com/astral-sh/ruff

### CODE REVIEW

**Review by hartwork — COMMENTED** [review]
Hi @jairhenrique I am wondering:
- whether ruff should be replacing flake8 (rather than adding to it) when https://beta.ruff.rs/docs/faq/#how-does-ruff-compare-to-flake8 says that only a "subset of the E and W rules (which originate from pycodestyle)" are implemented
- whether it's a gain or a loss to drop the `noqa` markers, in particular those with a number.

What do you think?

**Comment by jairhenrique** [comment]
> * whether it's a gain or a loss to drop the `noqa` markers, in particular those with a number.

I usually don't use the numbers, I can't tell if this is a loss or gain.

**Comment by hartwork** [comment]
@jairhenrique what the numbers allow in theory is to annotate why something was marked noqa semantically (i.e. documentation) but also to limit the scope of the noqa marker to just that aspect so that the QA tool won't silence unrelated issues now or in the future by accident.

I wonder, if we drop the markers, why doesn't ruff complain about the related cases, why doesn't ruff ask for the same issues to be intended and/or tolerable?

**Comment by jairhenrique** [comment]
@hartwork The ruff dropped `noqa` because I didn't enable the `E, F, W` linters.
Enabling these linters I needed to turn `noqa` back.

**Comment by jairhenrique** [comment]
@hartwork I returned the numbers in `noqa`. Removed `noqa` tags are from "fixed" issues.

File: pyproject.toml [hunk 1]
```diff
@@ -18,3 +18,22 @@ ignore-regex = "\\\\[fnrstv]"
...
+    "PIE", # flake8-pie
+    "RUF", # Ruff-specific rules
+    "UP",  # pyupgrade
+    "W",   # pycodestyle warning
+]
+line-length = 110
+target-version = "py38"
+
+[tool.ruff.isort]
+known-first-party = [ "vcr" ]
```
Comment (hartwork) [inline pyproject.toml:39]: A trailing newline would be nice here

File: pyproject.toml [hunk 2]
```diff
@@ -18,3 +18,22 @@ ignore-regex = "\\\\[fnrstv]"
...
+    "ISC", # flake8-implicit-str-concat
+    "PIE", # flake8-pie
+    "RUF", # Ruff-specific rules
+    "UP",  # pyupgrade
+    "W",   # pycodestyle warning
+]
+line-length = 110
+target-version = "py38"
+
+[tool.ruff.isort]
```
Comment (hartwork) [inline pyproject.toml:38]: I wonder what to do about the existing block `tool.isort` further up:
```ini
[tool.isort]
line_length = 110
known_first_party = "vcrpy"
multi_line_output = 3
use_parentheses = true
include_trailing_comma = true
```

In particular:
- does ruff read any of that? (assuming: no)
- should it be dropped? (assuming: yes)
- can any of these options ported to ruff more?

File: tests/integration/test_aiohttp.py [hunk 3]
```diff
@@ -278,9 +279,7 @@ def test_redirect(tmpdir, mockbin):
     # looking request_info.
     assert cassette_response.request_info.url == response.request_info.url
     assert cassette_response.request_info.method == response.request_info.method
-    assert {k: v for k, v in cassette_response.request_info.headers.items()} == {
-        k: v for k, v in response.request_info.headers.items()
-    }
+    assert dict(cassette_response.request_info.headers.items()) == dict(response.request_info.headers.items())
```
[the remark this answers is not in the document]
Comment (jairhenrique) [reply tests/integration/test_aiohttp.py]: This is from `ruff --fix .`.

File: tests/integration/test_httpx.py [hunk 4]
```diff
@@ -185,9 +187,7 @@ def test_redirect(mockbin, yml, do_request):
     # looking request_info.
     assert cassette_response.request.url == response.request.url
     assert cassette_response.request.method == response.request.method
-    assert {k: v for k, v in cassette_response.request.headers.items()} == {
-        k: v for k, v in response.request.headers.items()
-    }
+    assert dict(cassette_response.request.headers.items()) == dict(response.request.headers.items())
```
[the remark this answers is not in the document]
Comment (jairhenrique) [reply tests/integration/test_httpx.py]: This is from `ruff --fix .`.

File: vcr/__init__.py [hunk 5]
```diff
@@ -2,7 +2,7 @@
 from logging import NullHandler
 
 from .config import VCR
-from .record_mode import RecordMode as mode  # noqa import is not used in this file
+from .record_mode import RecordMode as mode  # noqa
```
[the remark this answers is not in the document]
Comment (jairhenrique) [reply vcr/__init__.py]: Done!

File: tests/integration/test_httpx.py [hunk 6]
```diff
@@ -2,10 +2,12 @@
 
 import pytest
 
+import vcr
+
 asyncio = pytest.importorskip("asyncio")
 httpx = pytest.importorskip("httpx")
 
-import vcr  # noqa: E402
+
```
[the remark this answers is not in the document]
Comment (jairhenrique) [reply tests/integration/test_httpx.py]: No! Fixed!

File: pyproject.toml (hunk 1 above)
Comment (jairhenrique) [reply pyproject.toml]: Done!

File: pyproject.toml (hunk 2 above)
Comment (jairhenrique) [reply pyproject.toml:31]: > does ruff read any of that?

No 😄 

> should it be dropped?

Yes, done!

> can any of these options ported to ruff more?

That's the default ruff configuration, but some settings configurations was not provided yet.

**Review by hartwork — APPROVED** [review]
@jairhenrique thanks for the adjustments! I have checked locally that flake8 does not report any errors against the latest state (except the now expected "E501 line too long" because flake8 no longer knows we want 110 columns). The Python changes are great and if for some reason ruff should turn out to have serious drawbacks over flake8/isort/pyflakes latter, nothing is blocking a cheap return to these tools, so the risk is minimal. Good work! :+1:

**Pull request merged**